From 493ff51f454f847962cacb0e92cb3d65b908e189 Mon Sep 17 00:00:00 2001
From: John Cox <jc@kynesim.co.uk>
Date: Sun, 11 Jun 2023 14:13:24 +0000
Subject: [PATCH 10/41] avcodec: Fake an initial timestamp if none received

This fixes playback of some .avi files that are missing PTSs


Upstream-Status: Inappropriate

RPI-Distro repo forks original vlc and applies patches

to enable raspiberry pi support.
---

 modules/codec/avcodec/video.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

--- a/modules/codec/avcodec/video.c
+++ b/modules/codec/avcodec/video.c
@@ -62,6 +62,9 @@ struct decoder_sys_t
     /* Video decoder specific part */
     date_t  pts;
 
+    mtime_t dts0;
+    bool dts0_used;
+
     /* Closed captions for decoders */
     cc_data_t cc;
 
@@ -585,6 +588,8 @@ static int InitVideoDecCommon( decoder_t
     /* ***** misc init ***** */
     date_Init(&p_sys->pts, 1, 30001);
     date_Set(&p_sys->pts, VLC_TICK_INVALID);
+    p_sys->dts0 = VLC_TICK_INVALID;
+    p_sys->dts0_used = false;
     p_sys->b_first_frame = true;
     p_sys->i_late_frames = 0;
     p_sys->b_from_preroll = false;
@@ -848,6 +853,8 @@ static void Flush( decoder_t *p_dec )
     AVCodecContext *p_context = p_sys->p_context;
 
     date_Set(&p_sys->pts, VLC_TICK_INVALID); /* To make sure we recover properly */
+    p_sys->dts0 = VLC_TICK_INVALID;
+    p_sys->dts0_used = false;
     p_sys->i_late_frames = 0;
     p_sys->b_draining = false;
     cc_Flush( &p_sys->cc );
@@ -875,6 +882,8 @@ static bool check_block_validity( decode
     if( block->i_flags & (BLOCK_FLAG_DISCONTINUITY|BLOCK_FLAG_CORRUPTED) )
     {
         date_Set( &p_sys->pts, VLC_TICK_INVALID ); /* To make sure we recover properly */
+        p_sys->dts0 = VLC_TICK_INVALID;
+        p_sys->dts0_used = false;
         cc_Flush( &p_sys->cc );
 
         p_sys->i_late_frames = 0;
@@ -1220,6 +1229,10 @@ static picture_t *DecodeBlock( decoder_t
             }
             if( b_has_data )
             {
+                /* Remember 1st DTS in case we need to invent a timebase */
+                if (p_sys->dts0 <= VLC_TICK_INVALID)
+                    p_sys->dts0 = p_block->i_dts;
+
                 pkt->data = p_block->p_buffer;
                 pkt->size = p_block->i_buffer;
                 pkt->pts = p_block->i_pts > VLC_TICK_INVALID ? p_block->i_pts : AV_NOPTS_VALUE;
@@ -1326,6 +1339,17 @@ static picture_t *DecodeBlock( decoder_t
         if( i_pts == AV_NOPTS_VALUE )
             i_pts = date_Get( &p_sys->pts );
 
+        /* VLC doesn't like having no pts - but a simple timestamp at the
+         * start of time is all that is needed to get it going - pick the
+         * first dts we saw as being in the right general area */
+        if (i_pts > VLC_TICK_INVALID)
+            p_sys->dts0_used = true;
+        else if (p_sys->dts0 > VLC_TICK_INVALID && !p_sys->dts0_used)
+        {
+            i_pts = p_sys->dts0;
+            p_sys->dts0_used = true;
+        }
+
         /* Interpolate the next PTS */
         if( i_pts > VLC_TICK_INVALID )
             date_Set( &p_sys->pts, i_pts );
